django/db/backends/mysql/introspection.py

```diff
@@ -202,13 +203,15 @@ def get_constraints(self, cursor, table_name):
             cursor.execute(type_query, [table_name])
             for constraint, check_clause in cursor.fetchall():
                 # Parse columns.
-                columns = OrderedSet()
-                for statement in sqlparse.parse(check_clause):
-                    for token in statement.flatten():
-                        if token.ttype in [sqlparse.tokens.Name, sqlparse.tokens.Literal.String.Single]:
-                            columns.add(token.value[1:-1])
+                cols = OrderedSet()
+                statement = sqlparse.parse(check_clause)[0]
+                tokens = (token for token in statement.flatten() if not token.is_whitespace)
+                for token in tokens:
+                    if token.ttype in (sqlparse.tokens.Name, sqlparse.tokens.Literal.String.Single):
+                        if token.value[1:-1] in columns:
+                            cols.add(token.value[1:-1])
```

Comment:
Can you explain why we don't use the same logic as in [the SQLite introspection](https://github.com/django/django/blame/c7bef16a749ccbe2dbc8ce6b697dc80146584384/django/db/backends/sqlite3/introspection.py#L286-L291), i.e.
- we use `sqlparse.tokens.Literal.String.Single` instead of `sqlparse.tokens.Literal.String.Symbol`,
- we don't use `sqlparse.tokens.Keyword`, and
- we unquote `sqlparse.tokens.Name`.

:thinking: 

I think we should add MySQL/MariaDB-specific tests of check constraints introspection (see [`test_check_constraint`](https://github.com/django/django/commit/782d85b6dfa191e67c0f1d572641d8236c79174c#diff-4379f72f6de99e27a38abec7669669a3R105-R117)).

Reply:
> Can you explain why we don't use the same logic as in [the SQLite introspection](https://github.com/django/django/blame/c7bef16a749ccbe2dbc8ce6b697dc80146584384/django/db/backends/sqlite3/introspection.py#L286-L291), i.e.

If we use the same logic as `SQLite introspection`, [introspection.tests.IntrospectionTests.test_get_constraints](https://github.com/django/django/blob/c7bef16a749ccbe2dbc8ce6b697dc80146584384/tests/introspection/tests.py#L215) will fail because when we run the test ``` `up_votes` >= 0 ``` will added to `information_schema.check_constraints` table.
When we parse this entry we have these 3 tokens:
- ``` `up_votes` ``` with type `Token.Name`
- `>=` with type `Token.Operator.Comparison`
- `0` with type `Token.Literal.Number.Integer`

So, ``` `up_votes` ``` will fall into the first block:
```
    if token.ttype in (sqlparse.tokens.Name, sqlparse.tokens.Keyword):
        if token.value in columns:
            cols.append(token.value)
```
and in this block we check `if token.value in columns`, and `token.value` is ``` `up_votes` ```.

I think it would be good to replace `token.value[1:-1]` with ```token.value.strip('`')``` to handle the check constraint like `JSON_VALID`.

About the token types, I just used these types (`tokens.Name`, `tokens.Literal.String.Single`)
because they are matched with our test cases.

Reply:
It's hard to say without tests. IMO we should add `_parse_constraint_columns(sql)` hook and various tests in the `tests/backends/mysql/test_introspection.py`.

Reply:
I will do it.